### PR TITLE
Cygwin compatibility

### PR DESCRIPTION
**Pull request by stukowski**
This pull request contains the necessary changes to make the KIM API compatible with the Cygwin environment on the Windows platform.  

The changes address the following issues:

- Windows looks for DLLs only in the current working directory and in directories listed in the ``PATH`` environment variable. 
- The GNU GCC compiler for Cygwin is set to strict ANSI mode and non-ISO POSIX functions are not available by default. This can be overcome by [undefining the ``__STRICT_ANSI__`` macro](https://stackoverflow.com/questions/21689124/mkstemp-and-fdopen-in-cygwin-1-7-28).
- DLLs should be installed in the ``/bin`` directory, not ``/lib``.
- ``CMAKE_SHARED_LIBRARY_PREFIX`` has the value ``"cyg"`` in the Cygwin environment. However ``"lib"`` should be used instead when generating the pkg-config file, ensuring a consistent naming across platforms. 

Only modifications to the CMake scripts but no source code changes were necessary. I've made sure that the CMake test suite runs without errors. Furthermore, I've built LAMMPS against the KIM API and successfully ran the example simulation scripts that use only a standard KIM model.

Please check the proposed changes for any potential compatibility problems on Unix/Linux platforms.

### CODE REVIEW

File: cpp/src/CMakeLists.txt [hunk 1]
```diff
@@ -100,3 +100,13 @@ target_include_directories(kim-loader PUBLIC
   ${CMAKE_BINARY_DIR}
   )
 target_link_libraries(kim-loader ${CMAKE_DL_LIBS})
+
+if(${CMAKE_SYSTEM_NAME} STREQUAL CYGWIN)
+  # The KIM API code makes use of non-standard POSIX functions such as mkstemp(), which are
```
Comment (ellio167) [inline cpp/src/CMakeLists.txt:105]: Do you have a list of the functions that are non-standard POSIX?  I would prefer to explore a switch to POSIX compliant code, if possible.

**Comment by stukowski** [comment]
I dug deeper into the POSIX function issue and have to admit that I was mislead somewhat by wrong information found on Stack Overflow and other places. Here is the updated picture I got of the situation:

The problematic functions used by the KIM API code are ``strdup()``, ``fdopen()``, the ``P_tmpdir`` macro, and maybe a few more. They *are* all standard functions, either POSIX or SVID, so I think using them is fine in general. However, they are not part of the ISO C standard and are thus not available when compiling the code in strict ANSI mode. Strict ANSI mode is activated by the ``-ansi`` or ``--std=c++98`` compiler flags, the latter being automatically set by CMake on the Cygwin platform.

According to the [glibc documentation](https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html) the POSIX and SVID functions can be made available even in strict ANSI mode by defining the ``_DEFAULT_SOURCE`` macro at compile time. This is my new solution. 

I don't think using these functions is generally wrong. They are part of the standards (just not the very limited ISO C standard) and they are our best option in terms of portability, it seems. There simply isn't anything more portable to replace them with (unless we decide to rely on C++17). Nevertheless, the solution proposed in this pull request is only preliminary in the sense that it ensures compatibility with Cygwin. Porting the KIM API to the Mingw-w64 compiler might require another approach.

**Comment by ellio167** [comment]
@stukowski Thanks for digging deeper.  This seems a very good explanation and solution.  (I'm also happy to hear that I, indeed, used standard compliant methods; I have tired to be careful to only use such things...)

**Comment by ellio167** [comment]
@stukowski Can you update this PR to have the "Allow edits from maintainers" option turned on?  I have some clean up and slight modifications that I would like to suggest/push.

**Comment by stukowski** [comment]
> @stukowski Can you update this PR to have the "Allow edits from maintainers" option turned on? I have some clean up and slight modifications that I would like to suggest/push.

The "Allow edits from maintainers" option was already turned on from the very beginning. Doesn't GitHub let you make modifications?

**Comment by ellio167** [comment]
OK, yes it's working; I must have done something wrong the first try.

I did a white space cleanup (trailing space removed and adding newline at end of file).

I also used the `PATH` update approach for the `shared_library_test` tests.  Since these tests could be run after the kim-api is installed (by compiling model drivers, etc. separately and using the installed kim-api) it seems undesirable to execute these tests with the `WORKINGDIRECTORY` set to the installed kim-api location (typically a protected system directory).

Let me know if you have any comments or objections.

**Comment by stukowski** [comment]
I agree, using the ``PATH``-based approach everywhere makes sense. However, your change broke the tests and I had to fix it with another commit.

Before, the string ``"shared-library-test"`` was used as executable target for ``add_test()``. Since ``shared-library-test`` happens to be the name of a CMake target, CMake automatically replaced the string with the full path to the executable when running the tests. Now that test execution is preceded by the ``PATH`` setup step, the automatic replacement doesn't work anymore. We have to specify the full path to the ``shared-library-test`` executable explicitly. I've done so by using a generator expression.

**Comment by ellio167** [comment]
Nice! Thanks for the fix.

**Pull request merged**